src/page-storage/store-page.js

```diff
@@ -1,66 +1,97 @@
-    return {finalPage}
+    return finalPage
 }
 
-// Tries to reidentify the page in the tab, or creates a new page doc for it.
+// Tries to reidentify the page in tab, or creates a new page doc for it.
 // Either way, it returns the page doc. If a new page was created, it starts
-// page analysis and then deduplication, and also returns a promise
-// (finalPagePromise) of the page doc resulting from these steps.
-export async function reidentifyOrStorePage({tabId, url}) {
+// page analysis and then deduplication, and also returns an async function
+// (fetchFinalPage) returning the page doc resulting from these steps.
+export async function reidentifyOrStorePage({ tabId, url }) {
     // Find pages we know that had the same URL.
-    const samePageCandidates = (await findPagesByUrl({url})).rows.map(row => row.doc)
+    const samePageCandidates = (await findPagesByUrl({ url })).rows.map(row => row.doc)
 
     // Check if we can already tell in advance that this is the same page.
-    const reusablePage = await tryReidentifyPage({tabId, url, samePageCandidates})
+    const reusablePage = await tryReidentifyPage({ tabId, url, samePageCandidates })
 
     if (reusablePage) {
         // Equality is known (or assumed) in advance. Reuse the old page as is.
-        return {page: reusablePage}
+        return { page: reusablePage }
     } else {
-        // Create a new page doc in the database.
-        const page = await createPageStub({url})
-
-        // Start analysis and (possibly) deduplication, but do not wait for it.
-        const finalPagePromise = analyseAndTryDedupePage({
-            tabId,
-            page,
-            samePageCandidates,
-        })
-
-        // Return the page stub, and a promise of the analysed & deduped page.
-        return {page, finalPagePromise}
+        // Create a new page doc stub in the database.
+        const page = await storePage(createPage({ url }))
+
+        // Set up fetching function to run anaylsis + fetching logic on-demand
+        const fetchFinalPage = () => analyseAndTryDedupePage({ tabId, page, samePageCandidates })
+
+        // Return the page stub, and a function to get the analysed & deduped page.
+        return { page, fetchFinalPage }
+    }
+}
+
+/**
+ * Given a URL, attempts to perform all logic needed to store a full page document.
+ * @param {string} url The URL to attempt to store a page document for.
+ * @returns {any} Page document that was created (or found, if reusable/deduped).
+ */
+export async function storePageFromUrl(url) {
+    // Find pages we know that had the same URL.
+    const samePageCandidates = (await findPagesByUrl({ url })).rows.map(row => row.doc)
+
+    // Check if we can already tell in advance that this is the same page.
+    const reusablePage = await tryReidentifyPage({ url, samePageCandidates })
+
+    // Equality is known (or assumed) in advance. Reuse the old page as is.
+    if (reusablePage) {
+        return reusablePage
     }
```

Comment:
Can we reuse `reindetifyOrStorePage` instead of duplicating the above lines?
It seems this whole function follows the same pattern as in the case storing a page from a tab; I suspect we should be able to make a single pipeline with a different implementation of the actual (fetch &) analysis (which may internally again use the same extraction functions). I made a [sketch](http://penpub.ink/page47.html) of how things fit in my mind. It does not answer the question how the data is stored. I like to have each finished piece of analysis be stored to database directly, for which it is easiest to store them inside the analysis function instead of returning the completed page object when ready. This is the current approach. It would also be clean if the page-analysis module does not know about the database, but it would then be hard to still incrementally store the analysis parts. Trade-offs..